model/versioning/ChangeSetItem.php

```diff
@@ -345,4 +345,41 @@ public static function get_for_object_by_id($objectID, $objectClass) {
 			'ObjectClass' => ClassInfo::baseDataClass($objectClass)
 		]);
 	}
+
+	/**
+	 * Gets the list of modes this record can be previewed in.
+	 *
+	 * @return array Map of mode names to url
+	 */
+	public function getPreviewLinks() {
+		$links = [];
+
+		// Preview draft
+		$stage = $this->getObjectInStage(Versioned::DRAFT);
+		if($stage instanceof CMSPreviewable && $stage->canView() && ($link = $stage->PreviewLink())) {
+			$links[Versioned::DRAFT] = Controller::join_links($link, '?stage=' . Versioned::DRAFT);
+		}
+
+		// Preview live
+		$live = $this->getObjectInStage(Versioned::LIVE);
+		if($live instanceof CMSPreviewable && $live->canView() && ($link = $live->PreviewLink())) {
+			$links[Versioned::LIVE] = Controller::join_links($link, '?stage=' . Versioned::LIVE);
+		}
+
+		return $links;
+	}
+
+	/**
+	 * Get edit link for this item
+	 *
+	 * @return string
+	 */
+	public function CMSEditLink()
+	{
+		$link = $this->getObjectInStage(Versioned::DRAFT);
+		if($link instanceof CMSPreviewable && $link->canEdit()) {
```

Comment:
Should this check canEdit, or just stick the link in anyway, and then trigger the "you can't access this item, please log in with a password that can" dialog? Actually, canView might be better anyway (you'd get a read-only form).